feat(curriculum): add nth-fibonacci number generator lab (#56026)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-nth-fibonacci-number-generator/66d9af3897e7d75a895b72c2.md

---
id: 66d9af3897e7d75a895b72c2
title: Build the nth Fibonacci number generator
challengeType: 14
dashedName: lab-nth-fibonacci-number-generator

> 

Reviewer: ```suggestion
```

---

# --description--

> 

Reviewer: ```suggestion

In mathematics, a sequence is an ordered collection of numbers. The Fibonacci sequence is a sequence of integer numbers in which each number is the sum of the two preceding numbers.

Here are the first ten numbers from the Fibonacci sequence: 0, 1, 1, 2, 3, 5, 8, 13, 21, 34.

```
I would move part of a user story here as description.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should define an array named `sequence` in the global scope. It should contain the number `0` and `1`.
1. You should create a function named `fibonacci`.
1. The `fibonacci` function should take a single parameter, which is a positive integer.

> 1. A call to `fibonacci(n)` should use a dynamic programming approach to compute and return the `n`-th number from the Fibonacci sequence, in which each number is the sum of the two preceding numbers.

Reviewer: ```suggestion
1. A call to `fibonacci(n)` should use a dynamic programming approach to compute and return the `n`-th number from the Fibonacci sequence.
```
Following the previous comment

1. Each computed number at the position `n` in the Fibonacci sequence should be stored at the index `n - 1` of the `sequence` array.

> 1. If the input is not a positive integer, the function should return "Invalid input".

Reviewer: ```suggestion
1. If the input is not a positive integer, the function should return `"Invalid input"`.
```

---
id: 66d9af3897e7d75a895b72c2
title: Build the nth Fibonacci number generator
challengeType: 14
dashedName: lab-nth-fibonacci-number-generator
---

# --description--

In this lab you will use a dynamic programming approach to compute and return the `n`-th number from the Fibonacci sequence, in which each number is the sum of the two preceding numbers.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should define an array named `sequence` in the global scope. It should contain the number `0` and `1`.
1. You should create a function named `fibonacci`.
1. The `fibonacci` function should take a single parameter, which is a positive integer.

> 1. If the input is not a positive integer, the function should return `"Invalid input"`.

Reviewer 2: The code also checks if the type is not of type number, this should be described in the user story.

> 1. The function should compute numbers in the Fibonacci sequence as needed.

Reviewer: ```suggestion
1. A call to `fibonacci(n)` should use a dynamic programming approach to compute and return the `n`-th number from the Fibonacci sequence.
```
In my opinion we should keep this as a user story and modify the description in some way. "The function should compute numbers in the Fibonacci sequence as needed" is a bit cryptic and we need to specify what the function should return.

1. Each computed number at the position `n` in the Fibonacci sequence should be stored at the index `n - 1` of the `sequence` array.

> 1. If the input number requests an alrady computed number, no new computations should happen and the number should be retrieved from the `sequence` array.

Reviewer: ```suggestion
1. If the input number requests an already computed number, no new computations should happen and the number should be retrieved from the `sequence` array.
```

# --hints--

You should have a `sequence` array.

```js
assert.isArray(sequence);
```

The `sequence` array should contain `0` and `1` at definition.

```js
assert.match(__helpers.removeJSComments(code), /(let|const)\s+sequence\s*=\s*\[\s*0\s*,\s*1\s*\]\s*;?/)
```

You should create a `fibonacci` function.

```js
assert.isFunction(fibonacci);
```

If the `fibonacci` function receives an argument that is not a positive integer it should return `"Invalid input"`.

```js
assert.strictEqual(fibonacci(3.5), "Invalid input")

> assert.strictEqual(fibonacci(0), "Invalid input")

Reviewer 2: Zero is not a positive nor a negative number. We should remove this or use a different input to test.

Reviewer: I was about to say the same, but the user story is accurate because it says "If the input is not a positive integer" so everything else is excluded. So I think this can be kept as is.
One thing that I would clarify is that tests are checking that the type is not different from "number".

Reviewer 2: I sort of agree with you. While in fact it states "if the input is not a positive integer.". We should make clear to the camper that the input could be of any kind.

This will otherwise cause too many issues on the forum.

Reviewer 3: I can remove the string test, but I see no reason to remove the 0

Reviewer 2: Let me reiterate, I think it's fine, but the user has too little context to go on. Most of the time when completing coding challenges the user is expected to exclude any negative number (If the user story is phrased like this).

 We could write an extra hint that checks invalid values?

Like: The fibonacci function should not accept `3.5` as a valid input.

and: The fibonacci function should not accept `0` as a valid input.

Reviewer 3: divide it into multiple tests? shure

assert.strictEqual(fibonacci(-1), "Invalid input")
assert.strictEqual(fibonacci("three"), "Invalid input")
```

`fibonacci(1)` should return `0`.

```js
assert.strictEqual(fibonacci(1), 0)

```

`fibonacci(32)` should return `1346269`.

```js
assert.strictEqual(fibonacci(32), 1346269)

```

`fibonacci(5)` should return `3`.

```js
assert.strictEqual(fibonacci(5), 3);

```

`sequence` should contain numbers up to the highest call so far.

```js
sequence.length = 2
fibonacci(2)
assert.lengthOf(sequence, 2)
fibonacci(5)
assert.lengthOf(sequence, 5)
fibonacci(3)

> assert.lengthOf(sequence, 5)

Reviewer 2: How can two different values have the same output? This doesn't make sense.

Reviewer 2: I see, that's still a little confusing. Could we add an example for this?

Reviewer: Where would you add the example? Before user stories?

Reviewer 3: if the camper compute the Fibonacci sequence as requested, outside the function, it doesn't need an example. This is checking that if the number already exist in the sequence, then no new numbers are added to the sequence

Reviewer 2: Okay, I will resolve this one.

fibonacci(10)
assert.lengthOf(sequence, 10)
assert.deepEqual(sequence, [0, 1, 1, 2, 3, 5, 8, 13, 21, 34])
```

# --seed--

## --seed-contents--

```js

```

# --solutions--

```js

> const sequence = [0, 1];

Reviewer 2: Sequence should be part of the function.

Reviewer: The idea is to have it outside the function and look at what has been already computed before.
If you call `fibonacci(n)`, then when you want to know the `m`-th element in the sequence you should get it directly from `sequence` if `m <= n` instead of computing it again.

---
id: 66d9af3897e7d75a895b72c2
title: Build the nth Fibonacci number generator
challengeType: 14
dashedName: lab-nth-fibonacci-number-generator
---

# --description--

In this lab you will use a dynamic programming approach to compute and return the `n`-th number from the Fibonacci sequence, in which each number is the sum of the two preceding numbers.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should define an array named `sequence` in the global scope. It should contain the number `0` and `1`.
1. You should create a function named `fibonacci`.
1. The `fibonacci` function should take a single parameter, which is a positive integer.

> 1. If the input is not a positive integer, the function should return `"Invalid input"`.

Reviewer 4: It should be mentioned in the instruction to check for other invalid inputs like a string as there is a test present for it,

```suggestion
1. If the input is a string or not a positive integer, the function should return `"Invalid input"`.
```

Reviewer: I think it's fine to keep it as is: the test has been split and one can clearly see that a string is not accepted as argument. Modifying the sentence in this way can cause more confusion than benefits in my opinion.
